File: Python/ASP Advanced/Graphs/Dijkstra's/Leetcode-1514.py
import heapq

class Solution(object):
    def maxProbability(self, n, edges, succProb, start_node, end_node):
        adj = {i:[] for i in range(n)}

        for i, edge in enumerate(edges):
            adj[edge[0]].append((succProb[i], edge[1]))
            adj[edge[1]].append((succProb[i], edge[0]))

        maxHeap = [(-1, start_node)]
        greatest = {i:0 for i in range(n)}
        while maxHeap and greatest[end_node] == 0:
            val, src = heapq.heappop(maxHeap)
            val = -val

            if greatest[src] != 0:
                continue

            greatest[src] = val

            for w, dst in adj[src]:
                if greatest[dst] == 0:
                    heapq.heappush(maxHeap, (-val*w, dst))

        return greatest[end_node]

sol = Solution()
n = 3
edges = [[0,1],[1,2],[0,2]]
succProb = [0.5,0.5,0.2]
start_node = 0
end_node = 2
print(sol.maxProbability(n, edges, succProb, start_node, end_node))

File: C++/ASP Advanced/Graphs/Dijkstra's/Leetcode-1514.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        unordered_map<int, vector<pair<double,int>>> adj;

        for (int i = 0; i < edges.size(); i++){
            vector<int> edge = edges[i];
            int a = edge[0], b = edge[1];
            double sp = succProb[i];
            adj[a].push_back({sp, b});
            adj[b].push_back({sp, a});
        }

        priority_queue<pair<double, int>> maxHeap;
        maxHeap.push({1, start_node});

        unordered_map<int, double> shortest;

        while (!maxHeap.empty()){
            auto [p1, src] = maxHeap.top();
            maxHeap.pop();

            if (shortest.count(src))continue;
            if (src == end_node)return p1;
            shortest[src] = p1;

            for (auto [p2, dst]: adj[src]){
                if (shortest.count(dst))continue;
                maxHeap.push({p1 * p2, dst});
            }
        }
        return 0;
    }
};